x/wasm/internal/keeper/query_plugins.go

```diff
@@ -146,6 +164,71 @@ func NoCustomQuerier(sdk.Context, json.RawMessage) ([]byte, error) {
 	return nil, wasmvmtypes.UnsupportedRequest{Kind: "custom"}
 }
 
+func IBCQuerier(wasm *Keeper, channelKeeper types.ChannelKeeper) func(ctx sdk.Context, caller sdk.AccAddress, request *wasmvmtypes.IBCQuery) ([]byte, error) {
+	return func(ctx sdk.Context, caller sdk.AccAddress, request *wasmvmtypes.IBCQuery) ([]byte, error) {
+		if request.PortID != nil {
+			contractInfo := wasm.GetContractInfo(ctx, caller)
+			res := wasmvmtypes.PortIDResponse{
+				PortID: contractInfo.IBCPortID,
+			}
+			return json.Marshal(res)
+		}
+		// I just discovered we cannot query by port (how odd since they store it like this)
+		// Anyway, I will allow Port as an optional filter
+		// TODO: let's revisit this query (if needed, if we should change it)
+		if request.ListChannels != nil {
+			portID := request.ListChannels.PortID
+			var channels wasmvmtypes.IBCEndpoints
+			channelKeeper.IterateChannels(ctx, func(ch types.IdentifiedChannel) bool {
+				if portID == "" || portID == ch.PortId {
+					newChan := wasmvmtypes.IBCEndpoint{
+						PortID:    ch.PortId,
+						ChannelID: ch.ChannelId,
+					}
+					channels = append(channels, newChan)
+				}
+				return false
+			})
+			res := wasmvmtypes.ListChannelsResponse{
+				Channels: channels,
+			}
+			return json.Marshal(res)
+		}
+		if request.Channel != nil {
+			channelID := request.Channel.ChannelID
+			_ = channelID
+			portID := request.Channel.PortID
+			if portID == "" {
+				contractInfo := wasm.GetContractInfo(ctx, caller)
+				portID = contractInfo.IBCPortID
+			}
+			got, found := channelKeeper.GetChannel(ctx, portID, channelID)
+			var channel *wasmvmtypes.IBCChannel
+			if found {
+				channel = &wasmvmtypes.IBCChannel{
+					Endpoint: wasmvmtypes.IBCEndpoint{
+						PortID:    portID,
+						ChannelID: channelID,
+					},
+					CounterpartyEndpoint: wasmvmtypes.IBCEndpoint{
+						PortID:    got.Counterparty.PortId,
+						ChannelID: got.Counterparty.ChannelId,
+					},
+					Order:               got.Ordering.String(),
+					Version:             got.Version,
+					CounterpartyVersion: got.Version, // FIXME: maybe we change the type? This data is not stored AFAIK
```

Comment:
ja, it is not stored with channel anymore. Too bad. We need the field during IBC handshake only. Otherwise it is empty.

Reply:
Shall I remove this from CosmWasm return value?

Reply:
`CounterpartyVersion` should be empty string here to be consistent with what we do during IBC handshake when the data is not available. The field has value during the handshake.